ros/docker/pure_pursuit/src/global_ros_pp.py

```diff
@@ -0,0 +1,191 @@
+        self.ygoal = 0
+    
+    def car_update(self,msg):
+        self.car.prev_xpos = self.car.xpos
+        self.car.xpos = msg.pose.pose.position.x
+        self.car.ypos = msg.pose.pose.position.y
+        self.car.yaw = msg.pose.pose.orientation.z
+        self.car.vel = msg.twist.twist.linear.x
+        
+    def set_waypoints(self,msg):
+        waypoints[:,0] = msg.PoseStamped[:].pose.position.x
+        waypoints[:,1] = msg.PoseStamped[:].pose.position.y
+        self.xgoal = msg.PoseStamped[N].pose.position.x
+        self.ygoal = msg.PoseStamped[N].pose.position.y
+
+# Publisher class that will publish Twist messages as the output of the controller
+class Publisher:
+    def __init__(self):
+        self.cmd = rospy.Publisher('/pure_pursuit/ppc_cmd', Float32MultiArray, queue_size=2)
+        self.rate = rospy.Rate(100)
+    
+    def publish(self,msg):
+        self.cmd.publish(msg)
+        self.rate.sleep()
+
+class NoIntersectionException(Exception):
+    def __init__(self):
+        self.message = 'no intersection'
+
+# Car class uses Odometry data to update the current position, yaw, and velocity of the car
+class Car():
+    def __init__(self, xpos=0, prev_xpos=0, ypos=0, yaw=0, vel=0):
+        self.xpos = xpos
+        self.prev_xpos = prev_xpos
+        self.ypos = ypos
+        self.yaw = yaw
+        self.vel = vel
+
+#--------------------------------------------(Steering Angle)--------------------------------------#
+
+# Calculates distance between the current position and a point along the path
+# Returns the index of the nearest waypoint
+def nearest_point(curr_x, curr_y):
+    dist = np.zeros(n_points)
+    dist = np.sqrt(((waypoints[:,0] - curr_x)**2) + ((waypoints[:,1] - curr_y)**2))
+    return np.argmin(dist)
+
+# Calculates the lookahead point based on the current position and the nearest waypoint 
+def lookahead(curr_x, curr_y):
+    position = np.array([curr_x, curr_y]).T 
+    wp = nearest_point(curr_x, curr_y)
+    try:
+        f = np.subtract(waypoints[wp] , position)
+        d = np.subtract(waypoints[wp+1] , waypoints[wp])   
+
+        a = np.dot(d,d); 
+        b = 2*np.dot(f,d); 
+        c = np.dot(f,f) - ld*ld
+        discriminant = (b*b) - (4*a*c) 
+        
+        if (discriminant<0):
+            raise NoIntersectionException
+        else:
+            discriminant = np.sqrt(discriminant)
+            t1 = (-b- discriminant)/(2*a)
+            t2 = (-b + discriminant)/(2*a)
+            if (t1>=0 and t1<=1.0):
+                point = waypoints[wp] + t1*d
+                pointx = point[0]
+                pointy = point[1]
+                return pointx, pointy
+                 #x and y value of ld
+            elif (t2 >= 0 and t2 <=1.0):
+                point = waypoints[wp] + t2*d
+                pointx = point[0]
+                pointy = point[1]
+                return pointx, pointy
+            else:
+                return 0,0
+
+    except NoIntersectionException as e:
```

Comment:
An exception is thrown and then immediately caught, so is there a point? In other words, what do you actually want to do when we have no intersection?